geonode/assets/utils.py

```diff
@@ -167,3 +201,50 @@ def rollback_asset_and_link(asset, link):
             asset.delete()  # TODO: make sure we are only deleting from DB and not also the stored data
     except Exception as e:
         logger.error(f"Could not rollback asset[{asset}] and link[{link}]", exc_info=e)
+
+
+def unlink_asset(resource: ResourceBase, asset: Asset, remove_asset: bool = True):
+    """
+    Unlinks an asset from a resource. By default, the asset is deleted.
+
+    Behavior:
+    - If the asset title is "Original" or "Files", it will not be unlinked or deleted.
+    - If remove_asset=False:
+        * Only the link between the resource and asset is removed.
+    - If remove_asset=True(default):
+        * If the asset is linked to other resources, only the link is removed.
+        * If the asset is only linked to the provided resource, the asset itself is deleted,
+          which in turn triggers the deletion of the associated files.
+    """
+    if asset.title in ["Original", "Files"]:
+        logger.info(f"Asset '{asset.title}' is protected and will not be unlinked or deleted.")
+        return False, "Asset is protected and will not be unlinked or deleted."
+
+    link = Link.objects.filter(resource=resource, asset=asset).first()
+    if not link:
+        return False, "Link between resource and asset not found."
+
+    if not remove_asset:
+        # Always just remove the link
+        link.delete()
+        msg = f"Asset {asset.pk} was unlinked from resource {resource.pk} but not deleted."
+        logger.info(msg)
+        return True, msg
+
+    # Check if the asset is linked to other resources
+    other_links_count = Link.objects.filter(asset=asset).exclude(pk=link.pk).count()
+
+    if other_links_count > 0:
+        # Only delete the link
+        link.delete()
+        msg = f"Asset {asset.pk} was unlinked but could not be deleted."
```

Comment:
The "could not" seems there are some unknown reasons. Let's log why it was not removed
```
msg = f"Asset {asset.pk} was unlinked but not deleted, because still linked to {other_links_count} resources"
```